Add ssdeep table for posix

sanevova opened the pull request:
This adds a new `ssdeep` table to darwin capable of computing fuzzy ssdeep hashes of files
[ssdeep home page](https://ssdeep-project.github.io/ssdeep/index.html)
This has been requested previously in https://github.com/facebook/osquery/issues/1775

Adding as a separate table instead of adding a column to `hash` table in the interest of performance since all columns are computed right now even if some are not SELECTed.

Leaks test:
```
$ ./tools/analysis/profile.py --leaks --query "select * from ssdeep where path = '/tmp/file' or directory = '/tmp/dir';"                                                                                                                                                                                                                               
Analyzing leaks in query: select * from ssdeep where path = '/tmp/file' or directory = '/tmp/dir';                                                                                                                                                                                                                                                                          
  definitely: 0 leaks for 0 total leaked bytes.                
```

theopolis commented:
Awesome work! We can trigger the tests as soon as the CLA is signed.

I have a question about the table vs column in the hash table choice. Is there something we can do to make this feasible as a column in hash? I think that would be most useful for people.

kcreyts commented:
Also why just in Darwin?

+1 to having it as a column in the cross-platform`file` table

On Fri, Jun 22, 2018 at 15:00 Teddy Reed <notifications@github.com> wrote:

> Awesome work! We can trigger the tests as soon as the CLA is signed.
>
> I have a question about the table vs column in the hash table choice. Is
> there something we can do to make this feasible as a column in hash? I
> think that would be most useful for people.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/osquery/pull/4629#issuecomment-399593950>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4Y6Fl8XREKnWqcd0gEM0UJd3r10moIks5t_WjhgaJpZM4U0YiT>
> .
>
-- 
Kyle Creyts

akindyakov commented on osquery/tables/system/darwin/ssdeep.cpp:
```diff
@@ -0,0 +1,100 @@
+                      QueryData& results) {
+  auto fd = fopen(path.c_str(), "r");
+  char *file_ssdeep_c_str = (char *)malloc(FUZZY_MAX_RESULT);
+  fuzzy_hash_file(fd, file_ssdeep_c_str);
```
`fuzzy_*` functions could return error (non-zero return value). So, to avoid any kind of incorrect data in the table it is better to handle `fuzzy_*` errors here. What do you think?

akindyakov commented on osquery/tables/system/darwin/ssdeep.cpp:
```diff
@@ -0,0 +1,100 @@
+                      const std::string& dir,
+                      QueryData& results) {
+  auto fd = fopen(path.c_str(), "r");
+  char *file_ssdeep_c_str = (char *)malloc(FUZZY_MAX_RESULT);
```
You don't need to allocate raw data on the stack and copy it then into `std::string` object. You could use `std::string` object from the beginning.
```c++
auto file_ssdepp_hash = std::string(FUZZY_MAX_RESULT, '\0');
fuzzy_hash_filename(path.c_str(),  &file_ssdepp_hash.front());
str.resize(str.find('\0', 0, str.size()));
```

akindyakov commented on osquery/tables/system/darwin/ssdeep.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+void genSsdeepForFile(const std::string& path,
+                      const std::string& dir,
```
You could use `.c_str ()` from `boost::filesystem :: path`, so you do not need to convert the boost path to the ` std::string` objects here.

akindyakov commented on osquery/tables/system/darwin/ssdeep.cpp:
```diff
@@ -0,0 +1,100 @@
+    LIKE,
+    paths,
+    ([&](const std::string& pattern, std::set<std::string>& out) {
+      std::vector<std::string> patterns;
```
I think `patterns` is not quite a good name for results of file pattern resolving.

akindyakov commented on osquery/tables/system/darwin/ssdeep.cpp:
```diff
@@ -0,0 +1,100 @@
+namespace tables {
+
+void expandFSPathConstraints(QueryContext& context,
+                             std::string path_column_name,
```
Looks like unused variable, isn't it?

akindyakov commented on osquery/tables/system/darwin/ssdeep.h:
```diff
@@ -0,0 +1,27 @@
+                      const std::string& dir,
+                      QueryData& results);
+
+QueryData genSsdeep(QueryContext& context);
```
It would be great to have a tests for this functionality. To avoid any bugs and involuntary failure in the future. What do you think?

akindyakov commented on specs/darwin/ssdeeep.table:
```diff
@@ -0,0 +1,12 @@
+description("Filesystem ssdeep data.")
+schema([
+    Column("path", TEXT, "Must provide a path or directory", index=True, required=True),
+    Column("directory", TEXT, "Must provide a path or directory", required=True),
```
Just a quick question. What the difference between `path` and `directory` columns ? And why one column with full path is not enough?

akindyakov commented:
Thank you for the contribution!

sanevova commented:
Thanks for the review, moved to `hash` table, moved dependency to POSIX

```
$ ./tools/analysis/profile.py --leaks --query "select * from hash where path = '/tmp/file' or directory = '/tmp/dir'"                                                                                                                                                                                                                                  
Analyzing leaks in query: select * from hash where path = '/tmp/file' or directory = '/tmp/dir'                                                                                                                                                                                                                                                                             
  definitely: 0 leaks for 0 total leaked bytes.
```

theopolis commented:
We’re still waiting for the CLA before we can test or merge the code. Please see the details/instructions from the bot’s post above :)

guliashvili commented on osquery/tables/system/tests/system_tables_tests.cpp:
```diff
@@ -277,6 +283,7 @@ TEST_F(HashTableTest, hashes_are_correct) {
   EXPECT_EQ(rows[0].at("md5"), contentMd5);
   EXPECT_EQ(rows[0].at("sha1"), contentSha1);
   EXPECT_EQ(rows[0].at("sha256"), contentSha256);
+  EXPECT_EQ(rows[0].at("ssdeep"), contentSsdeep);
```
If it's not posix system you don't query ssdeep column. Do you?

guliashvili commented on osquery/tables/system/hash.cpp:
```diff
@@ -210,11 +221,36 @@ void genHashForFile(const std::string& path,
+}
+
+void expandFSPathConstraints(QueryContext& context,
+                             std::string path_column_name,
```
I think, const std::string& path_column_name should be enough.

guliashvili commented on osquery/tables/system/hash.cpp:
```diff
@@ -182,6 +186,13 @@ bool FileHashCache::load(const std::string& path, MultiHashes& out) {
+  std::string file_ssdeep_hash(FUZZY_MAX_RESULT, '\0');
+  auto did_ssdeep_fail =
+    fuzzy_hash_filename(path.c_str(), &file_ssdeep_hash.front());
+  return did_ssdeep_fail ? "-1" : std::move(file_ssdeep_hash);
```
Is it normal ssdeep to fail? I think we need to log it.

guliashvili commented:
ok to test

guliashvili commented on osquery/tables/system/hash.cpp:
```diff
@@ -210,11 +225,36 @@ void genHashForFile(const std::string& path,
+  results.push_back(std::move(r));
+}
+
+void expandFSPathConstraints(QueryContext& context,
```
👍

guliashvili commented on osquery/tables/system/hash.cpp:
```diff
@@ -182,6 +186,21 @@ bool FileHashCache::load(const std::string& path, MultiHashes& out) {
 }
 
+std::string genSsdeepForFile(const std::string& path) {
+#ifndef WINDOWS
```
ssdeep looks like the feature every OS could benefit. Is it possible to implement it for windows also?

sanevova replied:
I agree that Windows can benefit from it.
But since osquery dependency management is completely different for Windows AFAIK, I suggest that this is done in a separate PR.

guliashvili commented:
Overall, to me, it looks very good feature.
I think query writers would benefit a lot if we also supported stored procedure computing edit distance between the hashes. (I think ssdeep also implements it)

akindyakov commented on osquery/tables/system/hash.cpp:
```diff
@@ -182,6 +186,21 @@ bool FileHashCache::load(const std::string& path, MultiHashes& out) {
 }
 
+std::string genSsdeepForFile(const std::string& path) {
+#ifndef WINDOWS
```
May be it is better to use `#ifdef OSQUERY_POSIX` ? It is not a big deal until `not WINDOWS == OSQUERY_POSIX`. But it may be changed one day.

akindyakov commented on osquery/tables/system/hash.cpp:
```diff
@@ -182,6 +186,21 @@ bool FileHashCache::load(const std::string& path, MultiHashes& out) {
+    LOG(WARNING) << "ssdeep failed: " << path;
+    return "-1";
+  }
+  return file_ssdeep_hash;
```
According to https://ssdeep-project.github.io/ssdeep/doc/api/html/fuzzy_8h.html output could be less than FUZZY_MAX_RESULT. To prevent any misunderstanding and even bug it is better to trim the null-bytes-tail.
```c++
str.resize(str.find('\0', 0, str.size()));
```

akindyakov commented on osquery/tables/system/hash.cpp:
```diff
@@ -182,6 +186,21 @@ bool FileHashCache::load(const std::string& path, MultiHashes& out) {
+      fuzzy_hash_filename(path.c_str(), &file_ssdeep_hash.front());
+  if (did_ssdeep_fail) {
+    LOG(WARNING) << "ssdeep failed: " << path;
+    return "-1";
```
As far as `"-1"` is a default value for this column, it it better to define it as a named constant and use constant by name in such places. What do you think?

akindyakov commented:
FreeBSD build fails with error 
```
[ 75%] Building CXX object osquery/tables/CMakeFiles/osquery_system_tables.dir/system/hash.cpp.o
/var/jenkins/workspace/osqueryPullRequestFreeBSD/TargetSystem/aws-freebsd-11/osquery/tables/system/hash.cpp:21:10: fatal error: 'fuzzy.h' file not found
#include <fuzzy.h>
```
looks like there is something wrong with dependency installation in FreeBSD.
